src/Interpreters/Cache/LRUFileCachePriority.cpp

```diff
@@ -370,25 +374,40 @@ void LRUFileCachePriority::LRUIterator::invalidate()
         "Invalidating entry in LRU queue. Key: {}, offset: {}, previous size: {}",
         entry->key, entry->offset, entry->size);
 
+    chassert(entry->size != 0);
     cache_priority->updateSize(-entry->size);
     cache_priority->updateElementsCount(-1);
     entry->size = 0;
 }
 
-void LRUFileCachePriority::LRUIterator::updateSize(int64_t size)
+void LRUFileCachePriority::LRUIterator::incrementSize(size_t size, const CacheGuard::Lock &)
 {
     assertValid();
 
     const auto & entry = *iterator;
     LOG_TEST(
         cache_priority->log,
-        "Update size with {} in LRU queue for key: {}, offset: {}, previous size: {}",
+        "Increment size with {} in LRU queue for key: {}, offset: {}, previous size: {}",
         size, entry->key, entry->offset, entry->size);
 
     cache_priority->updateSize(size);
     entry->size += size;
 }
 
+void LRUFileCachePriority::LRUIterator::decrementSize(size_t size)
+{
+    assertValid();
+
+    const auto & entry = *iterator;
+    LOG_TEST(
+        cache_priority->log,
+        "Decrement size with {} in LRU queue for key: {}, offset: {}, previous size: {}",
+        size, entry->key, entry->offset, entry->size);
+
+    cache_priority->updateSize(-size);
+    entry->size -= size;
```

Comment:
(I'm not familiar with that part of code base)
1. Since `decrementSize` is called without a lock, what guarantees that there is no data race, e.g. if `cache_priority` is accessed by several parallel threads?
2. I think there should be an assert that entry->size >= size

Reply:
> what guarantees that there is no data race

`entry->size` is `std::atomic<size_t>`.

> if cache_priority is accessed by several parallel threads

This is not possible for `cache_priority` by itself. It is always protected by `cache_lock`. But fields of a struct `IFileCachePriority::Entry` (an element in `cache_priority`, which is a shared pointer and is stored not only in `main_priority`) can be accessed by different threads in parallel. Specifically, only `size` field (which is atomic) in this struct is changed with or without lock. And to be more precise - `size` is always incremented under lock, but can be decremented without lock. The former is done without lock to avoid extra contention (this was not introduced in this PR, this behaviour exists for a long time) and because it does not break the guarantees (e.g. does not break the logic of `tryReserve` method, it is ok that at some point cache size decreases while we are reserving space - we will just break earlier). 

> I think there should be an assert that entry->size >= size


Good point, will add.